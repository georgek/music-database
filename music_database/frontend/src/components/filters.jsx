import React, { useEffect, useRef, useState } from "react";

import Button from "react-bootstrap/Button";
import Card from "react-bootstrap/Card";
import Dropdown from "react-bootstrap/Dropdown";
import InputGroup from "react-bootstrap/InputGroup";
import Form from "react-bootstrap/Form";
import FormControl from "react-bootstrap/FormControl";

export function SearchBox(props) {
  const input = useRef(null);
  useEffect(() => {
    const timeoutID = setTimeout(() => input.current.focus(), 100);
    return () => clearTimeout(timeoutID);
  }, []);
  return (
    <Form.Group>
      <Form.Label>{props.label}</Form.Label>
      <InputGroup className="mb-3">
        <FormControl
          ref={input}
          name={props.name}
          value={props.value || ""}
          onChange={e => props.onChange(e.target.value)}
          placeholder={props.placeholder}
          aria-label={props.label}
          aria-describedby="basic-addon2"
        />
        <Button variant="danger" onClick={props.onRemove}>
          Remove
        </Button>
      </InputGroup>
    </Form.Group>
  );
}

export function ChoiceBox(props) {
  const [choices, setChoices] = useState([]);

  async function fetchChoices(url) {
    const response = await fetch(url);
    const data = await response.json();
    setChoices(data.results);
  }

  useEffect(() => {
    fetchChoices(props.choicesUrl);
  }, [props.choicesUrl]);

  return (
    <Form.Group>
      <Form.Label>{props.label}</Form.Label>
      <InputGroup className="mb-3">
        <FormControl
          as="select"
          name={props.name}
          value={props.value}
          onChange={e => props.onChange(e.target.value)}
          aria-label={props.label}
          aria-describedby="basic-addon2"
        >
          <option value="">(All)</option>
          {choices.map(choice => (
            <option key={choice.id} value={choice.id}>
              {choice.name}
            </option>
          ))}
        </FormControl>
        <Button variant="danger" onClick={props.onRemove}>
          Remove
        </Button>
      </InputGroup>
    </Form.Group>
  );
}

export default function FilterSet(props) {
  const [activeFilters, setActiveFilters] = useState(props.activeFilters);
  const [timerId, setTimerId] = useState(null);

  useEffect(() => {
    setActiveFilters(props.activeFilters);
  }, [props.activeFilters]);

  const filters = Object.fromEntries(
    props.availableFilters.map(filter => [filter.key, filter])
  );

  function updateFilters(filterKey, value) {
    const newFilters = Object.assign({}, activeFilters);
    newFilters[filterKey] = value;
    setActiveFilters(newFilters);
    if (timerId) {
      clearTimeout(timerId);
    }
    setTimerId(
      setTimeout(() => {
        props.onFiltersChange(newFilters);
        setTimerId(null);
      }, 200)
    );
  }

  function addFilter(filterKey) {
    const newFilters = Object.assign({}, activeFilters);
    newFilters[filterKey] = "";
    setActiveFilters(newFilters);
    props.onAddFilter(filterKey);
  }

  function removeFilter(filterKey) {
    const newFilters = Object.assign({}, activeFilters);
    delete newFilters[filterKey];
    setActiveFilters(newFilters);
    props.onRemoveFilter(filterKey);
  }

  return (
    <Card body>
      <Card.Title>Filters</Card.Title>
      <Form>
        {activeFilters &&
          Object.keys(activeFilters).map(filterKey => {
            let filter = filters[filterKey];
            switch (filter.type) {
              case "search":
                return (
                  <SearchBox
                    key={filter.key}
                    name={filter.key}
                    label={filter.name}
                    value={activeFilters[filter.key]}
                    placeholder={`Filter on ${filter.name}`}
                    onChange={value => updateFilters(filterKey, value)}
                    onRemove={() => removeFilter(filterKey)}
                  />
                );
              case "choice":
                return (
                  <ChoiceBox
                    key={filter.key}
                    name={filter.key}
                    label={filter.name}
                    choicesUrl={filter.choicesUrl}
                    value={activeFilters[filter.key]}
                    onChange={value => updateFilters(filterKey, value)}
                    onRemove={() => removeFilter(filterKey)}
                  />
                );
              default:
                return null;
            }
          })}
      </Form>
      <Dropdown onSelect={key => addFilter(key)}>
        <Dropdown.Toggle variant="primary" id="addFilterDropdown">
          Add filter
        </Dropdown.Toggle>
        <Dropdown.Menu>
          {props.availableFilters.map(filter => {
            if (activeFilters && filter.key in activeFilters) {
              return null;
            } else {
              return (
                <Dropdown.Item key={filter.key} eventKey={filter.key}>
                  {filter.name}
                </Dropdown.Item>
              );
            }
          })}
        </Dropdown.Menu>
      </Dropdown>
    </Card>
  );
}
