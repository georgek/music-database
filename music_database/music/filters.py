from django_filters import rest_framework as filters

from .models import Genre, Track


class TrackTableFilter(filters.FilterSet):
    class Meta:
        model = Track
        fields = ["name", "album", "artist", "genre"]

    name = filters.CharFilter(field_name="name", lookup_expr="icontains")
    album = filters.CharFilter(field_name="album__title", lookup_expr="icontains")
    artist = filters.CharFilter(
        field_name="album__artist__name", lookup_expr="icontains"
    )
    genre = filters.ModelChoiceFilter(field_name="genre", queryset=Genre.objects.all())
