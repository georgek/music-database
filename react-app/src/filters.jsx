import React, { useState, useEffect } from "react";

import Button from "react-bootstrap/Button";
import Card from "react-bootstrap/Card";
import InputGroup from "react-bootstrap/InputGroup";
import Form from "react-bootstrap/Form";
import FormControl from "react-bootstrap/FormControl";

import "bootstrap/dist/css/bootstrap.min.css";
import "./index.css";

function SearchBox(props) {
  return (
    <Form.Group>
      <Form.Label>{props.label}</Form.Label>
      <InputGroup className="mb-3">
        <FormControl
          name={props.name}
          value={props.value}
          onChange={(e) => props.onChange(e.target.value)}
          placeholder={props.placeholder}
          aria-label={props.label}
          aria-describedby="basic-addon2"
        />
        {props.value &&
          <InputGroup.Append>
            <Button
              variant="outline-secondary"
              onClick={() => props.onChange("")}
            >
              Clear
            </Button>
          </InputGroup.Append>}
      </InputGroup>
    </Form.Group>
  );
}

function ChoiceBox(props) {
  const [choices, setChoices] = useState([]);

  useEffect(() => {
    console.log(props.choicesUrl);
    fetch(props.choicesUrl)
      .then(response => response.json())
      .then(data => setChoices(data.results));
  }, [props.choicesUrl]);

  return (
    <Form.Group>
      <Form.Label>{props.label}</Form.Label>
      <InputGroup className="mb-3">
        <FormControl
          as="select"
          name={props.name}
          value={props.value}
          onChange={(e) => props.onChange(e.target.value)}
          aria-label={props.label}
          aria-describedby="basic-addon2"
        >
          <option value="">(All)</option>
          {choices.map(
            (choice) =>
              <option key={choice.id} value={choice.id}>
                {choice.name}
              </option>
          )}
        </FormControl>
      </InputGroup>
    </Form.Group>
  );
}

export default class FilterSet extends React.Component {
  constructor(props) {
    super(props);

    this.filterKeys = props.filters.map(
      (field) => field.key
    );
    const filterEntries = this.filterKeys.map(
      (key) => [key, ""]
    );

    this.state = {
      filters: Object.fromEntries(filterEntries),
      searchString: "",
    };
    this.handleFiltersChange = this.handleFiltersChange.bind(this);
    this.handleSearchStringChange = this.handleSearchStringChange.bind(this);
  }

  handleFiltersChange(key, value) {
    const filters = Object.assign({}, this.state.filters);

    filters[key] = value;

    this.setState({
      filters: filters,
    });
    this.props.onFiltersChange(filters);
  }

  handleSearchStringChange(value) {
    this.setState({
      searchString: value,
    });
    this.props.onSearchStringChange(value);
  }

  render() {
    return (
      <Card body>
        <Card.Title>Filters</Card.Title>
        <Form>
          {this.props.filters.map(
            (field) => {
              switch (field.type) {
              case "search":
                return (
                  <SearchBox
                    key={field.key}
                    name={field.key}
                    label={field.name}
                    value={this.state.filters[field.key]}
                    placeholder={`Filter on ${field.name}`}
                    onChange={
                      (value) => this.handleFiltersChange(field.key, value)
                    }
                  />
                );
              case "choice":
                return (
                  <ChoiceBox
                    key={field.key}
                    name={field.key}
                    label={field.name}
                    choicesUrl={field.choicesUrl}
                    value={this.state.filters[field.key]}
                    onChange={
                      (value) => this.handleFiltersChange(field.key, value)
                    }
                  />
                );
              default:
                return null;
              }
            }
          )}
        </Form>
      </Card>
    );
  }
}
