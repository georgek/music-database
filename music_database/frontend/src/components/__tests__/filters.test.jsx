import React from "react";
import { SearchBox } from "../filters";
import renderer from "react-test-renderer";

describe("SearchBox", () => {
  it("renders correctly", () => {
    const component = renderer.create(
      <SearchBox
        name="name"
        label="label"
        value="value"
        placeholder="placeholder"
        onChange={() => null}
        onRemove={() => null}
      />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
