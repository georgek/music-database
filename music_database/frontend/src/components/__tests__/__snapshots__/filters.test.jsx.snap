// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBox renders correctly 1`] = `
<div
  className="form-group"
>
  <label
    className="form-label"
  >
    label
  </label>
  <div
    className="mb-3 input-group"
  >
    <input
      aria-describedby="basic-addon2"
      aria-label="label"
      className="form-control"
      name="name"
      onChange={[Function]}
      placeholder="placeholder"
      value="value"
    />
    <button
      className="btn btn-danger"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Remove
    </button>
  </div>
</div>
`;
